pkg/sql/catalog/hydratedtables/hydratedcache.go

```diff
@@ -0,0 +1,327 @@
+	}
+	metaMisses = metric.Metadata{
+		Name:        "sql.hydrated_table_cache.misses",
+		Help:        "counter on the number of cache misses",
+		Measurement: "reads",
+		Unit:        metric.Unit_COUNT,
+		MetricType:  io_prometheus_client.MetricType_COUNTER,
+	}
+)
+
+// CacheSize controls the size of the LRU cache.
+var CacheSize = settings.RegisterNonNegativeIntSetting(
+	"sql.catalog.hydrated_tables.cache_size",
+	"number of table descriptor versions retained in the hydratedtables LRU cache",
+	128)
+
+// NewCache constructs a new Cache.
+func NewCache(settings *cluster.Settings) *Cache {
+	c := &Cache{
+		settings: settings,
+		metrics:  makeMetrics(),
+	}
+	c.mu.cache = cache.NewOrderedCache(cache.Config{
+		Policy: cache.CacheLRU,
+		ShouldEvict: func(size int, key, value interface{}) bool {
+			return size > int(CacheSize.Get(&settings.SV))
+		},
+		OnEvicted: func(key, value interface{}) {
+			putCacheKey(key.(*cacheKey))
+		},
+	})
+	return c
+}
+
+type hydratedTableDescriptor struct {
+	tableDesc *tabledesc.Immutable
+	typeDescs []*cachedType
+}
+
+type cachedType struct {
+	name    tree.TypeName
+	typDesc catalog.TypeDescriptor
+}
+
+type cacheKey lease.IDVersion
+
+func (c cacheKey) Compare(comparable llrb.Comparable) int {
+	o := comparable.(*cacheKey)
+	switch {
+	case c.ID < o.ID:
+		return -1
+	case c.ID > o.ID:
+		return 1
+	default:
+		switch {
+		case c.Version < o.Version:
+			return -1
+		case c.Version > o.Version:
+			return 1
+		default:
+			return 0
+		}
+	}
+}
+
+var _ llrb.Comparable = (*cacheKey)(nil)
+
+var cacheKeySyncPool = sync.Pool{
+	New: func() interface{} { return new(cacheKey) },
+}
+
+func newCacheKey(table catalog.TableDescriptor) *cacheKey {
+	k := cacheKeySyncPool.Get().(*cacheKey)
+	*k = cacheKey{
+		ID:      table.GetID(),
+		Version: table.GetVersion(),
+	}
+	return k
+}
+
+func putCacheKey(k *cacheKey) {
+	*k = cacheKey{}
+	cacheKeySyncPool.Put(k)
+}
+
+// GetHydratedTableDescriptor returns an ImmutableCopy with the types
+// hydrated. It may use a cached copy but all of the relevant type descriptors
+// will be retrieved via the TypeDescriptorResolver.
+func (c *Cache) GetHydratedTableDescriptor(
+	ctx context.Context, table *tabledesc.Immutable, res catalog.TypeDescriptorResolver,
+) (_ *tabledesc.Immutable, canUseCached bool, err error) {
+	if table.IsModified() {
+		return nil, false, nil
+	}
+	if !table.ContainsUserDefinedTypes() {
+		return table, true, nil
+	}
+
+	// TODO(ajwerner): This cache may thrash a bit right when a version of a type
+	// changes as different callers oscillate evicting the old and new versions of
+	// that type. It should converge rapidly but nevertheless, a finer-granularity
+	// cache could mitigate the problem. The idea would be to cache all tuples
+	// of table and type versions and then check if what we get from the resolver
+	// matches any of them.
+	var groupKey string // used as a proxy for cache hit
+	defer func() {
+		if err != nil || !canUseCached {
```

Comment:
I'm not sure that this is going to work for you -- there are cases where the `err` variable is getting shadowed and won't set this. Should wrap the main logic in a function or lambda here and deal with the response

Reply:
It doesn’t matter if it’s in scope or shadowed or whatever. When you have a named return value it gets set to the return value in the return statement.

Reply:
Huh, TIL that is some strange behavior...

Reply:
I think of this as a pretty critical feature in the who `defer` scheme. It's very important to be able to rely on the idea that when you're running a defer you'll be able to see the values being returned from the function (unless they are modified by an earlier return). 

One handy trick is

```
func f(args interface{}) (ret interface{}, err error) {
    defer func() {  log.Infof(..., args, ret, err) }(args)
}
```